<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"> 
<html lang="en">
<head>
<title>CGView - Gallery</title>
<meta name="keywords" content="CGView" />
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" />
<link rel="stylesheet" href="styles/stylesheet.css"  type="text/css" />
</head>
<body>


<table class="main" border="0" cellspacing="6" cellpadding="2"> <!-- START MAIN TABLE -->
<tbody>

<tr> <!-- MAIN TABLE HEADER ROW -->
    <td class="header">
    <table class="header" width="100%" border="0" cellspacing="0" cellpadding="2"> <!-- START HEADER TABLE -->
    <tbody>

    <tr>
    <td class="header_title" valign="middle">
    <span class="title">CGView - Circular Genome Viewer</span><br />
    <span class="subtitle">Gallery</span><br />
    </td>
    </tr>

    <tr>
    <td class="link_bar" valign="middle">
<a href="index.html">What is CGView?</a> | <a href="gallery.html">Gallery</a> | <a href="create_overview.html">Creating a Map</a> | <a href="application.html">CGView Application</a> | <a href="api_overview.html">CGView API</a> | <a href="http://stothard.afns.ualberta.ca/cgview_server/">CGView Server</a>
    </td>
    </tr>

    </tbody>
    </table> <!-- END HEADER TABLE -->
    </td>
</tr>

<tr> <!-- MAIN TABLE CONTENT ROW -->


<td class="content" align="center" valign="top">  <!-- MAIN TABLE MAIN COLUMN -->

    <table class="content" border="0" cellspacing="4" cellpadding="2"> <!-- START CONTENT TABLE -->
    <tbody>
    <tr>  
    <td>
        <span class="subtitle">Gallery</span><br />
        Use the links on this page to view sample maps. See <a href="create_overview.html">creating CGView maps</a> for information on downloading and using CGView. Note that CGView can be used to display any type of feature information, from Gene Ontology classifications to sequence similarity scores.<br /><br />


        <span class="bold">Examples of browsable maps created by CGView:</span><br />
        
        <ul>
        <li>
        <a href="images/featureRange_element_series/index.html" target="_new">Example 1</a> - A very simple navigable map. Format: PNG and SVG.
        </li>
        </ul>

        <span class="bold">Examples of static images created by CGView:</span> <br />

        <ul>
        <li>
        <a href="gallery/halocaridina.html">Example 2 (337 kilobytes)</a> - Halocaridina rubra Holthuis mitochondrial genome (map courtesy of Scott Santos, Auburn University). Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/unknown_genome_2.html">Example 3 (955 kilobytes)</a> - Large map showing genes, COGs, GC content and GC skew. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/AE001273.html">Example 4 (1,424 kilobytes)</a> - Large map of Chlamydia trachomatis D/UW-3/CX. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/NC_000117_zoom.html">Example 5 (767 kilobytes)</a> - Large map of Chlamydia trachomatis D/UW-3/CX. Zoom: 10X zoom on base 100,000. Format: PNG.
        </li>
        <li>
        <a href="gallery/NC_003385.html">Example 6 (384 kilobytes)</a> - Salmonella enterica plasmid pHCM2. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/e_coli_svg.html">Example 7 (2 megabytes)</a> - Escherichia coli. Zoom: 6X zoom on base 4000000. Format: SVG.
        </li>
        <li>
        <a href="gallery/pblue_sk_svg.html">Example 8 (536 kilobytes)</a> - pBlueScript SK-. Zoom: unzoomed. Format: SVG.
        </li>
        <li>
        <a href="gallery/NC_003098.html">Example 9 (4,545 kilobytes)</a> - Large map of Streptococcus pneumoniae R6. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/NC_001823_svg.html">Example 10 (13 megabytes)</a> - Reclinomonas americana mitochondrion complete genome. Zoom: unzoomed. Format: SVG.
        </li>
        <li>
        <a href="gallery/AJ560763.html">Example 11 (601 kilobytes)</a> - Temperate bacteriophage Aaphi23 of Actinobacillus actinomycetemcomitans. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/sample_project_1c.html">Example 12 (814 kilobytes)</a> - Map showing BLAST comparison results between Aster yellows witches'-broom plasmid pAYWB-III and plasmids pAYWB-I, pAYWB-II, and pAYWB-IV. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/sample_project_5a.html">Example 13 (7.3 megabytes)</a> - Map showing BLAST comparison results between Reclinomonas americana mitochondrion and the E. coli genome, Arabidopsis mitochondrion, and human mitochondrion. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/NC_000909.html">Example 14 (18 megabytes)</a> - Large map showing BLAST comparison results for translated CDS features from Methanocaldococcus jannaschii DSM 2661 vs translated CDS features from 100 bacterial genomes. Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/montage.png">Example 15 (28 megabytes)</a> - Montage showing BLAST comparisons of translated CDS features from various E. coli plasmids. COGs are also shown, in the first and fourth tracks (counting from outside of circles). Zoom: unzoomed. Format: PNG.
        </li>
        <li>
        <a href="gallery/NC_004431_medium.png">Example 16 (8 megabytes)</a> - Map comparing E. coli CFT073 CDS translations to CDS translations from all E. coli chromosomes in RefSeq. COG categories and sequence features are shown in the outermost four rings.
        </li>
        <li>
        <a href="gallery/NC_004431_x-large.png">Example 17 (70 megabytes)</a> - Same as the previous example but drawn at a larger size (12000 x 12000, or 144 megapixels) and with gene labels. To view this image right-click the link and choose Save Link As (or some other similar option) and then open the image in a dedicated graphics program.
        </li>
        </ul>

        <p>Need a custom map? Contact Paul Stothard: stothard@ualberta.ca</p>

        <span class="bold">Citing CGView:</span><br />
        <ul>
        <li>
        <a href="https://pubmed.ncbi.nlm.nih.gov/15479716/">Stothard P, Wishart DS (2005) Circular genome visualization and exploration using CGView. Bioinformatics 21:537-539.</a>
        </li>
        </ul>

    </td>
    </tr>
    </tbody>
    </table> <!-- END CONTENT TABLE -->

</td>
</tr>

<tr> <!-- MAIN TABLE FOOTER ROW -->
    <td class="footer">
    <table class="footer" border="0" cellspacing="0" cellpadding="2"> <!-- START FOOTER TABLE -->
    <tbody>
    <tr>
    <td valign="middle">
    CGView was written by Paul Stothard. Contact: stothard@ualberta.ca
    </td>
    </tr>
    </tbody>
    </table> <!-- END FOOTER TABLE -->
    </td>
</tr>

<tr>
<td>
</td>
</tr>

</tbody>
</table> <!-- END MAIN TABLE -->
</body>
</html>
